presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestReorderWindows.java

```diff
@@ -0,0 +1,257 @@
+    private final Symbol shipdate;
+    private final Symbol receiptdate;
+
+    private final SymbolReference discountReference;
+    private final SymbolReference quantityReference;
+    private final SymbolReference taxReference;
+
+    public TestReorderWindows()
+    {
+        queryRunner = new LocalQueryRunner(testSessionBuilder()
+                .setCatalog("local")
+                .setSchema("tiny")
+                .build());
+
+        queryRunner.createCatalog(queryRunner.getDefaultSession().getCatalog().get(),
+                new TpchConnectorFactory(queryRunner.getNodeManager(), 1),
+                ImmutableMap.of());
+
+        commonFrame = Optional.empty();
+
+        suppkey = new Symbol("suppkey");
+        orderkey = new Symbol("orderkey");
+        partkey = new Symbol("partkey");
+        shipdate = new Symbol("shipdate");
+        receiptdate = new Symbol("receiptdate");
+        tax = new Symbol("tax");
+
+        discountReference = new SymbolReference("discount");
+        quantityReference = new SymbolReference("quantity");
+        taxReference = new SymbolReference("tax");
+
+        windowA = new WindowNode.Specification(
+                ImmutableList.of(suppkey),
+                ImmutableList.of(orderkey),
+                ImmutableMap.of(orderkey, SortOrder.ASC_NULLS_LAST));
+
+        windowAp = new WindowNode.Specification(
+                ImmutableList.of(suppkey),
+                ImmutableList.of(shipdate),
+                ImmutableMap.of(shipdate, SortOrder.ASC_NULLS_LAST));
+
+        windowApp = new WindowNode.Specification(
+                ImmutableList.of(suppkey, tax),
+                ImmutableList.of(receiptdate),
+                ImmutableMap.of(receiptdate, SortOrder.ASC_NULLS_LAST));
+
+        windowB = new WindowNode.Specification(
+                ImmutableList.of(partkey),
+                ImmutableList.of(receiptdate),
+                ImmutableMap.of(receiptdate, SortOrder.ASC_NULLS_LAST));
+    }
+
+    private static String getWindowDescription(WindowNode.Specification specification)
+    {
+        return format("partition by %s order by %s",
+                specification.getPartitionBy().stream().map(Symbol::getName).collect(Collectors.joining(", ")),
+                specification.getOrderBy().stream().map(orderSymbol -> sortItemToString(orderSymbol, specification.getOrderings())).collect(Collectors.joining(", ")));
+    }
+
+    private static String sortItemToString(Symbol sortItem, Map<Symbol, SortOrder> orderings)
+    {
+        SortOrder sortOrder = orderings.get(sortItem);
+        return format("%s %s %s", sortItem, (sortOrder.isAscending() ? "ASC" : "DESC"), (sortOrder.isNullsFirst() ? "NULLS FIRST" : "NULLS LAST"));
+    }
+
+    @Test
+    public void testNonMergeableABAReordersToAABAllOptimizers()
+    {
+        @Language("SQL") String sql = "select " +
+                "sum(quantity) over(" + getWindowDescription(windowA) + ") sum_quantity_A, " +
```

Comment:
Added some tests in https://github.com/Teradata/presto/pull/332/commits/b8f911edc22ec65e6fd3ee58fbe672568a4893ac. Will rebase it to be first test in this PR. Didn't want to rebase anything to not screw up commit hashes.